src/widgets/immunizations/immunizations-form.component.tsx

```diff
@@ -0,0 +1,342 @@
+import React, { useEffect, useRef, useState } from "react";
+import SummaryCard from "../../ui-components/cards/summary-card.component";
+import styles from "./immunizations-form.css";
+import { DataCaptureComponentProps } from "../shared-utils";
+import { useTranslation } from "react-i18next";
+import { savePatientImmunization } from "./immunizations.resource";
+import { mapToFHIRImmunizationResource } from "./immunization-mapper";
+import { useCurrentPatient } from "@openmrs/esm-api";
+import { useHistory } from "react-router-dom";
+import { createErrorHandler } from "@openmrs/esm-error-handling";
+import { getStartedVisit } from "../visit/visit-utils";
+import useSessionUser from "../../utils/use-session-user";
+import {
+  ImmunizationFormData,
+  ImmunizationSequence
+} from "./immunization-domain";
+
+export function ImmunizationsForm(props: ImmunizationsFormProps) {
+  const [vaccineName, setVaccineName] = useState("");
+  const [vaccineUuid, setVaccineUuid] = useState("");
+  const [immunizationObsUuid, setImmunizationObsUuid] = useState("");
+  const [vaccinationDate, setVaccinationDate] = useState(null);
+  const [immunizationSequences, setImmunizationSequences] = useState([]);
+  const [currentDose, setCurrentDose] = useState<ImmunizationSequence>(
+    {} as ImmunizationSequence
+  );
+  const [vaccinationExpiration, setVaccinationExpiration] = useState(null);
+  const [lotNumber, setLotNumber] = useState(undefined);
+  const [manufacturer, setManufacturer] = useState("");
+  const [enableCreateButtons, setEnableCreateButtons] = useState(false);
+  const [enableEditButtons, setEnableEditButtons] = useState(true);
+  const [viewEditForm, setViewEditForm] = useState(true);
+  const [formChanged, setFormChanged] = useState<Boolean>(false);
+  const [
+    isLoadingPatient,
+    patient,
+    patientUuid,
+    patientErr
+  ] = useCurrentPatient();
+  const formRef = useRef<HTMLFormElement>(null);
+  const { t } = useTranslation();
+  const history = useHistory();
+  const today = new Date().toISOString().split("T")[0];
+  const currentUser = useSessionUser();
+
+  useEffect(() => {
+    if (vaccinationDate) {
+      setEnableCreateButtons(true);
+    } else {
+      setEnableCreateButtons(false);
+    }
+  }, [vaccinationDate]);
+
+  useEffect(() => {
+    if (viewEditForm && formChanged) {
+      setEnableEditButtons(true);
+    } else {
+      setEnableEditButtons(false);
+    }
+  }, [viewEditForm, formChanged]);
+
+  useEffect(() => {
+    if (props.match.params) {
+      const {
+        immunizationObsUuid,
+        vaccineName,
+        vaccineUuid,
+        manufacturer,
+        expirationDate,
+        vaccinationDate,
+        lotNumber,
+        sequences,
+        currentDose
+      }: ImmunizationFormData = props.match.params;
+
+      setImmunizationObsUuid(immunizationObsUuid);
```

Comment:
All these are re-rendering. Sure we don't want a joint state for them? If you set multiple state entries in the same synchronous code its an indicator that the states are joined and should be treated as one and not independently. In these scenarios, e.g., `useReducer` shines.

Reply:
Hmm! We could do that. Though this is just one case (In case of edit when everything need to be used from previously saved data). I would still need to change the individual state for form field `onSelect` and `onChange`. 

If it is one single object(That's what I understand from joint state), I would need to call something like `setFormState({lotNumber:event.target.value}/setFormState({date:event.target.value}` etc every-time. Is this what you are suggesting.

Reply:
Yes in that direction.

So its either like one object that sets its sub-entries via, e.g.,

```ts
const [formState, setFormState] = React.useState({});
const updateSingle = <T extends keyof typeof formState>(name: T, value: (typeof formState)[T]) => setFormState(state => ({
  ...state,
  [name]: value,
}));
```

or alternatively use the `useReducer`, which gives you this flexibility in a more structured way.

Reply:
@FlorianRappl I have made this change please have a look! I have used the method `updateSingle`.
Thanks